Add some more methods that exist for the matlab TestCase class

Sbte opened the pull request:
I added an interface so you can call the assertions on the TestCase classes like in Matlab. So for instance you can now do

```
function test_case(t)
t.assertEqual(1,1)
```

I also implemented 4 additional methods that can be called in this way. I spent only a few hours on this, so there might be some errors, especially in the documentation, since I mostly copy-pasted that, but I wanted to make sure if you are actually interested in merging these changes before looking further into that. The code itself should be mostly correct since I added a lot of tests.

Also, I'm not sure if there's any naming convention I should stick to for the errors that I throw.

Sbte commented:
The decreased coverage can mostly be explained by Matlab-only code

nno commented on MOxUnit/assertGreaterThan.m:
```diff
@@ -0,0 +1,73 @@
+%   'moxunit:differentClass         a and b are of different class
+%   'moxunit:differentSparsity'     a is sparse and b is not, or
+%                                         vice versa
+%   'moxunit:elementsNotEqual'      values in a and b are not equal
```
This needs updating, it's probably from a copy-paste .

nno commented on MOxUnit/assertGreaterThan.m:
```diff
@@ -0,0 +1,73 @@
+function assertGreaterThan(a, b, message)
+% assert that a is greater than b
+%
+% assertGreaterThan(a,b,[msg])
```
Question: is it intended that a and b must be scalars? This would make sense to me, but it would be good to (1) document this explicitly and (2) ensure that the code reflects this. 

Other question: is it intended to require that a and b are numeric?

Sbte replied:
I'll have a look on Monday to see what Matlab does in different cases so we can do something similar. It could be that they apply any or all if arrays are passed. And it would be nice to have tests succeed that also succeed in Matlab. I'll also fix the comments related to this.

nno commented:
Thanks a lot. It would be fine with me to add the less-than and greater-than assert functions, I just added a few small questions and a note about a documentation detail. I would be happy to hear your thoughts about these. 
Generally this looks like good work and I appreciate your efforts, in particular all the tests that were added. So I think we're not so far off from merging into master.

Sbte commented:
What about the warning assert function? To me this is the most important one since you can not mimic this by an assert true, unlike the other two functions. And I personally use it a lot.

ThomasFeher commented:
I'm not sure it is a good idea to copy-paste from Matlab (maybe even documentation) due to possible copyright issues. Just mentioning.

Sbte commented:
Don't worry, I'm not going to copy-paste anything from Matlab. The documentation was copy-pasted from the MOxUnit assertEqual method. As for the implementation, it will be different from Matlab since I will not be looking at their code, but it would be nice to have (somewhat) similar behavior. The behavior itself can just be obtained through tests.

nno commented:
Thanks @ThomasFeher and @Sbte for making explicit that we should not copy-paste anything from Matlab. As @Sbte pointed out, wherever necessary the behaviour of Matlab functions can be deduced from tests.

Sbte commented:
It seems like Matlab mostly just does whatever `<` does with an `any` around it, so I removed the size check. This is handled internally anyway by both Octave and Matlab. I added a test for that. Note that this also works:

```
[1,2] < [3;4]
ans =

  1  1
  1  1
```

nno commented on MOxUnit/assertGreaterThan.m:
```diff
@@ -0,0 +1,54 @@
+        whatswrong='inputs are not of the same class';
+        error_id='assertGreaterThan:classNotEqual';
+
+    elseif length(a) == 1 && length(b) == 1 && a <= b
```
NaNs behave special so this doesn't work for them I think; instead of ``a<=b`` use ``~(a > b)``

nno commented on MOxUnit/assertGreaterThan.m:
```diff
@@ -0,0 +1,54 @@
+        whatswrong='first input argument in not larger than the second';
+        error_id='assertGreaterThan:notGreaterThan';
+
+    elseif any(a <= b)
```
- see previous NaN comment
- we may have to vectorize, in case a and b are not vectors but (e.g.) 4-dimensional.

nno commented on tests/test_assert_greater_than.m:
```diff
@@ -0,0 +1,20 @@
+    assertExceptionThrown(@()assertGreaterThan([1 2],[1,2,3]));
+    assertExceptionThrown(@()assertGreaterThan(1,1), ...
+                          'assertGreaterThan:notGreaterThan');
+    assertExceptionThrown(@()assertGreaterThan([2,3,3],[1,2,3]), ...
```
Can you please add tests with (1) scalar NaN compared with other NaN, scalar NaN compared with scalar non-NaN, (3) 4-dimensional NaN compared with 4-dimensional NaN, e.g. zeros(2,2,2,2)+NaN; (4) 4-dimensional arrays of the same size but different non-NaN values? Also for the less_than tests please.

nno commented on tests/test_assert_greater_than.m:
```diff
@@ -0,0 +1,20 @@
+                          'assertGreaterThan:notGreaterThan');
+
+function test_assert_greater_than_passes
+    assertGreaterThan(1,0);
```
Can you add a non-vector case?

nno commented on tests/test_assert_greater_than.m:
```diff
@@ -0,0 +1,20 @@
+function test_assert_greater_than_exceptions
+    assertExceptionThrown(@()assertGreaterThan([1],'a'), ...
+                          'assertGreaterThan:classNotEqual');
+    assertExceptionThrown(@()assertGreaterThan([1 2],[1,2,3]));
```
Add the exception type?

nno commented:
Thanks for the updates, I made some suggestions. In particular (1) correctly dealing with NaN, (2) testing n-dimensional arrays.

Sbte commented:
I don't understand your comments about the vectorization. I added tests for

```
assertGreaterThan([1,1],[0;0]);
```

so arrays that have dimensions 1x2 and 2x1, which should not fail since it gives back a 2x2 array, which is something that also works with Matlab and I also added a test with

```
assertExceptionThrown(@()assertGreaterThan([1 2],[1,2,3]));
```

which are vectors of actual different sizes, which can not be compared. This just gives back an internal Matlab or Octave error, which is also the same thing Matlab does. I can check the error values separately for both Octave and Matlab, but I am not sure what extra tests should be added.

nno commented:
I'm much concerned about the current behaviour that ``assertGreaterThan([1,1],[0;0]);`` does not raise an exception. I see it as an ugly consequence of broadcasting, which IIRC was introduced recently in Matlab (a few years ago at most). This could lead to cases of a test passing unexpectedly.

Instead I would strongly prefer to change the code so that ``assertGreaterThan`` always fails if the sizes of the inputs are not equal, just as ``assertEqual`` does. To keep it consistent, I would prefer to have the check for equal size /before/ the check for equal class. 

This should also address different exception ids for ``assertExceptionThrown(@()assertGreaterThan([1 2],[1,2,3]));`` in Matlab and Octave. 

What do you think?

Sbte commented:
Well, I'd prefer to have it most similar to Matlab, which means that it would accept `assertGreaterThan([1,1],[0;0]);` as valid, as well as maybe a more common use case, which is `assertGreaterThan([1,2,3,4;5,6,7,8],0);`, which means all values have to be greater than 0. You could argue that a user should instead do `assertGreaterThan(min([1,2,3,4;5,6,7,8]),0);`. So if you prefer to have the size check, I can add it. The big advantage of this would be that `assertGreaterThan([2,1],[1;0]);` would not return false when a user accidentally forgot to transpose one of the two input arguments.

As for the similarity with `assertEqual`: I think that this is a completely different method since this can also be used for many other objects. And the `Equal` implies that for arrays the sizes should also be equal.

Sbte commented:
I added all your suggestions.

nno commented:
@Sbte thanks for all the work, much appreciated!

nno commented:
@Sbte: I've added you to the Contributions section in the README file.

Sbte commented:
Thanks. Now I can update my own tests to fetch MOxUnit from this repository instead of my own fork ^^
